p2p/src/authenticated/lookup/actors/tracker/actor.rs

```diff
@@ -0,0 +1,427 @@
+use super::{
+    directory::{self, Directory},
+    ingress::{Mailbox, Message, Oracle},
+    Config,
+};
+use commonware_cryptography::Signer;
+use commonware_runtime::{Clock, Handle, Metrics as RuntimeMetrics, Spawner};
+use futures::{channel::mpsc, StreamExt};
+use governor::clock::Clock as GClock;
+use rand::Rng;
+use tracing::debug;
+
+/// The tracker actor that manages peer discovery and connection reservations.
+pub struct Actor<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> {
+    context: E,
+
+    // ---------- Configuration ----------
+    /// The maximum number of peers in a set.
+    max_peer_set_size: usize,
+
+    // ---------- Message-Passing ----------
+    /// The mailbox for the actor.
+    receiver: mpsc::Receiver<Message<E, C::PublicKey>>,
+
+    // ---------- State ----------
+    /// Tracks peer sets and peer connectivity information.
+    directory: Directory<E, C::PublicKey>,
+}
+
+impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> Actor<E, C> {
+    /// Create a new tracker [`Actor`] from the given `context` and `cfg`.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        context: E,
+        cfg: Config<C>,
+    ) -> (Self, Mailbox<E, C::PublicKey>, Oracle<E, C::PublicKey>) {
+        // Sign my own information
+        let myself = (cfg.crypto.public_key(), cfg.address);
+
+        // General initialization
+        let directory_cfg = directory::Config {
+            mailbox_size: cfg.mailbox_size,
+            max_sets: cfg.tracked_peer_sets,
+            rate_limit: cfg.allowed_connection_rate_per_peer,
+            allow_private_ips: cfg.allow_private_ips,
+        };
+        let directory = Directory::init(context.clone(), myself, directory_cfg);
+        let (sender, receiver) = mpsc::channel(cfg.mailbox_size);
+
+        (
+            Self {
+                context,
+                max_peer_set_size: cfg.max_peer_set_size,
+                receiver,
+                directory,
+            },
+            Mailbox::new(sender.clone()),
+            Oracle::new(sender),
+        )
+    }
+
+    /// Start the actor and run it in the background.
+    pub fn start(mut self) -> Handle<()> {
+        self.context.spawn_ref()(self.run())
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.receiver.next().await {
+            self.directory.process_releases();
```

Comment:
As discussed, we should update to: https://github.com/commonwarexyz/monorepo/pull/1108